File: boj/TopologicalSorting/BOJ_2056.py
# 백준 2056 작업
# 21.03.15
import sys
from functools import cmp_to_key
from collections import deque
from itertools import combinations
from itertools import permutations
rd = sys.stdin.readline
INF = int(1e9)

n = int(input())
time = [0] * (n + 1)
ans = [0] * (n + 1)
indegree = [0] * (n +1)
edges = [[] for _ in range(n + 1)]

for i in range(1, n + 1):
    temp = list(map(int, rd().split()))
    time[i] = ans[i] = temp[0]
    if temp[1] > 0:
        for j in temp[2:]:
            edges[j].append(i)
            indegree[i] += 1

q = []
for i in range(1, n + 1):
    if indegree[i] == 0:
        q.append(i)

while q:
    now = q.pop()

    for i in edges[now]:
        indegree[i] -= 1
        if ans[now] + time[i] > ans[i]: ans[i] = ans[now] + time[i]
        if indegree[i] == 0: q.append(i)

print(max(ans))

File: boj/TopologicalSorting/BOJ_2056.swift
// 백준 2056 작업
// 21.03.23

import Foundation

let n = Int(readLine()!)!

var graph = Array(repeating: Array<Int>(), count: n + 1)
var indegree = Array(repeating: 0, count: n + 1)
var time = Array(repeating: 0, count: n + 1)
var ans = Array(repeating: 0, count: n + 1)

for i in stride(from: 1, through: n, by: 1) {
  let arr = readLine()!.split(separator: " ").map { Int(String($0))! }
  time[i] = arr[0]
  
  for j in stride(from: 2, to: 2 + arr[1], by: 1) {
    graph[arr[j]].append(i)
    indegree[i] += 1
  }
}

var q: [Int] = []
for i in stride(from: 1, through: n, by: 1) {
  if indegree[i] == 0 {
    q.append(i)
    ans[i] = time[i]
  }
}

while !q.isEmpty {
  let now = q.removeLast()
  
  for i in graph[now] {
    indegree[i] -= 1
    ans[i] = max(ans[i], ans[now] + time[i])
    if indegree[i] == 0 { q.append(i) }
  }
}

print(ans.max()!)
